extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/indexing/MSQCompactionStrategy.java

```diff
@@ -0,0 +1,437 @@
+          ),
+          ColumnType.LONG,
+          new ExprMacroTable(Collections.singletonList(new TimestampFloorExprMacro()))
+      );
+      virtualColumns = VirtualColumns.create(virtualColumn);
+    }
+    return virtualColumns;
+  }
+
+  private static Query<?> buildGroupByQuery(CompactionTask compactionTask, Interval interval, DataSchema dataSchema)
+  {
+    DimFilter dimFilter = dataSchema.getTransformSpec().getFilter();
+
+    GroupByQuery.Builder builder = new GroupByQuery.Builder()
+        .setDataSource(new TableDataSource(compactionTask.getDataSource()))
+        .setVirtualColumns(getVirtualColumns(dataSchema))
+        .setDimFilter(dimFilter)
+        .setGranularity(new AllGranularity())
+        .setDimensions(getAggregateDimensions(dataSchema))
+        .setAggregatorSpecs(Arrays.asList(dataSchema.getAggregators()))
+        .setContext(compactionTask.getContext())
+        .setInterval(interval);
+
+    if (compactionTask.getTuningConfig() != null && compactionTask.getTuningConfig().getPartitionsSpec() != null) {
+      getOrderBySpec(compactionTask.getTuningConfig().getPartitionsSpec()).forEach(builder::addOrderByColumn);
+    }
+    return builder.build();
+  }
+
+  private Map<String, Object> createMSQTaskContext(CompactionTask compactionTask, DataSchema ds)
+      throws JsonProcessingException
+  {
+    Map<String, Object> context = new HashMap<>(compactionTask.getContext());
+    context.put(
+        DruidSqlInsert.SQL_INSERT_SEGMENT_GRANULARITY,
+        jsonMapper.writeValueAsString(ds.getGranularitySpec() != null
+                                      ? ds.getGranularitySpec()
+                                          .getSegmentGranularity()
+                                      : DEFAULT_SEGMENT_GRANULARITY)
+    );
+    if (!isQueryGranularityEmpty(ds)) {
+      context.put(
+          DruidSqlInsert.SQL_INSERT_QUERY_GRANULARITY,
+          jsonMapper.writeValueAsString(ds.getGranularitySpec().getQueryGranularity())
+      );
+    }
+    return context;
+  }
+
+  private static TaskStatus runSubtasks(
+      List<MSQControllerTask> tasks,
+      TaskToolbox toolbox,
+      CurrentSubTaskHolder currentSubTaskHolder,
+      String compactionTaskId
+  ) throws JsonProcessingException
+  {
+    final int totalNumSpecs = tasks.size();
+    log.info("Generated [%d] MSQControllerTask specs", totalNumSpecs);
+
+    int failCnt = 0;
+
+    for (MSQControllerTask eachTask : tasks) {
+      final String json = toolbox.getJsonMapper().writerWithDefaultPrettyPrinter().writeValueAsString(eachTask);
+      if (!currentSubTaskHolder.setTask(eachTask)) {
+        String errMsg = "Task was asked to stop. Finish as failed.";
+        log.info(errMsg);
+        return TaskStatus.failure(compactionTaskId, errMsg);
+      }
+      try {
+        if (eachTask.isReady(toolbox.getTaskActionClient())) {
+          log.info("Running MSQControllerTask: " + json);
+          final TaskStatus eachResult = eachTask.run(toolbox);
```

Comment:
Not sure if I understand it correctly, but I assume all controller task should go through `ControllerImpl` to start.
There is only controller task no worker task?